Step 1: Import libraries and load the NHS Prescription Cost Analysis (PCA) datasets

In this section, the required Python libraries are imported, and all monthly NHS PCA CSV files are identified and loaded into Python. Each monthly dataset is stored temporarily before being combined into a single dataframe. This creates one consolidated dataset that will be explored, cleaned, and analysed in the following steps.

Import required libraries

The libraries required for this project are imported. These libraries provide functionality for data manipulation, numerical operations, data visualisation, and locating multiple CSV files within the project folder.

1.1 Import libraries

In [21]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import glob
import os

Locate the monthly NHS datasets

The glob library is used to search the data folder and identify every CSV file. This allows all monthly datasets to be processed automatically without manually specifying each filename.

1.2 Find CSV files

In [24]:
# Find all CSV files stored in the data folder
files = glob.glob("../data/*.csv")

# Display the number of CSV files found
len(files)

12

1.3 Load each CSV

In [25]:
# Create an empty list to temporarily store each monthly dataframe
df_list = []

# Loop through each CSV file path found in the files list
for file in files:

    # Read the current CSV file and convert it into a pandas dataframe
    temp_df = pd.read_csv(file)

    # Add the monthly dataframe to the list
    df_list.append(temp_df)

Load the monthly datasets

Each CSV file is read into a pandas dataframe using pd.read_csv(). Every dataframe is then stored in a list, making it possible to combine all monthly datasets into a single dataframe in the next step.

1.4 Concatenate

In [27]:
# Combine all monthly dataframes into one large dataframe
# ignore_index=True creates a new continuous row index
df = pd.concat(df_list, ignore_index=True)

Combine the monthly datasets

The individual monthly dataframes are combined into one consolidated dataframe using pd.concat(). The ignore_index=True parameter creates a new sequential index, ensuring the combined dataset has continuous row numbering.

Step 2 - Inspect the combined dataset

Explore the combined NHS Prescription Cost Analysis (PCA) dataset to understand its structure, dimensions, data types, missing values, duplicate records, and overall data quality before performing any cleaning or analysis.

2.1 View first rows

In [28]:
# Display the first 5 rows of the combined dataset
# This helps us understand the data structure and values
df.head()

,YEAR_MONTH,REGION_NAME,REGION_CODE,ICB_NAME,ICB_CODE,DISPENSER_ACCOUNT_TYPE,BNF_PRESENTATION_CODE,BNF_PRESENTATION_NAME,SNOMED_CODE,SUPPLIER_NAME,...,BNF_SECTION_CODE,BNF_SECTION,BNF_CHAPTER_CODE,BNF_CHAPTER,PREP_CLASS,PRESCRIBED_PREP_CLASS,ITEMS,TOTAL_QUANTITY,NIC,PHARMACY_ADVANCED_SERVICE
0,202505,MIDLANDS,Y60,NHS NORTHAMPTONSHIRE INTEGRATED CARE BOARD,QPM,English Dispensing Doctor,010300000BBADAD,diabact UBT 50mg tablets,9.384211e+15,Mayoly UK Ltd,...,103,Antisecretory drugs and mucosal protectants,1,Gastro-Intestinal System,3,2,1,1.0,21.25,Not applicable
1,202505,EAST OF ENGLAND,Y61,NHS CAMBRIDGESHIRE AND PETERBOROUGH INTEGRATED...,QUE,English Pharmacy,010300000BBADAD,diabact UBT 50mg tablets,9.384211e+15,Mayoly UK Ltd,...,103,Antisecretory drugs and mucosal protectants,1,Gastro-Intestinal System,3,2,1,1.0,21.25,Not applicable
2,202505,NORTH WEST,Y62,NHS GREATER MANCHESTER INTEGRATED CARE BOARD,QOP,English Pharmacy,0103010D0AAABAB,Cimetidine 400mg tablets,4.235221e+16,NaN,...,103,Antisecretory drugs and mucosal protectants,1,Gastro-Intestinal System,1,1,315,14267.0,1320.12,Not applicable
3,202505,NORTH EAST AND YORKSHIRE,Y63,NHS HUMBER AND NORTH YORKSHIRE INTEGRATED CARE...,QOQ,English Pharmacy,0103010D0AAABAB,Cimetidine 400mg tablets,4.235221e+16,NaN,...,103,Antisecretory drugs and mucosal protectants,1,Gastro-Intestinal System,1,1,146,8061.0,745.76,Not applicable
4,202505,NORTH EAST AND YORKSHIRE,Y63,NHS SOUTH YORKSHIRE INTEGRATED CARE BOARD,QF7,English Pharmacy,0103010D0BBADAD,Tagamet 200mg/5ml syrup,7.727110e+14,Rosemont Pharmaceuticals Ltd,...,103,Antisecretory drugs and mucosal protectants,1,Gastro-Intestinal System,3,3,1,1200.0,56.98,Not applicable


Display the First Five Rows

The first five rows of the dataset are displayed to gain an initial understanding of the data structure, column contents, and formatting.

2.2 Check dataset size

In [29]:
# Display the number of rows and columns in the combined dataset
df.shape

(6876045, 27)

Check the Dataset Dimensions

The number of rows and columns in the combined dataset is displayed to confirm that all monthly datasets have been successfully combined.

2.3 Check column names

In [30]:
# Display all column names in the dataset
# This helps identify which columns are available for analysis
df.columns.tolist()

['YEAR_MONTH',
 'REGION_NAME',
 'REGION_CODE',
 'ICB_NAME',
 'ICB_CODE',
 'DISPENSER_ACCOUNT_TYPE',
 'BNF_PRESENTATION_CODE',
 'BNF_PRESENTATION_NAME',
 'SNOMED_CODE',
 'SUPPLIER_NAME',
 'UNIT_OF_MEASURE',
 'GENERIC_BNF_EQUIVALENT_CODE',
 'GENERIC_BNF_EQUIVALENT_NAME',
 'BNF_CHEMICAL_SUBSTANCE_CODE',
 'BNF_CHEMICAL_SUBSTANCE',
 'BNF_PARAGRAPH_CODE',
 'BNF_PARAGRAPH',
 'BNF_SECTION_CODE',
 'BNF_SECTION',
 'BNF_CHAPTER_CODE',
 'BNF_CHAPTER',
 'PREP_CLASS',
 'PRESCRIBED_PREP_CLASS',
 'ITEMS',
 'TOTAL_QUANTITY',
 'NIC',
 'PHARMACY_ADVANCED_SERVICE']

Display the Column Names

All column names are listed to understand the available variables and identify those that are relevant to the research questions.

2.4 Check data types and missing values

In [31]:
# Display information about columns, data types,
# and the number of non-null values
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6876045 entries, 0 to 6876044
Data columns (total 27 columns):
 #   Column                       Dtype  
---  ------                       -----  
 0   YEAR_MONTH                   int64  
 1   REGION_NAME                  str    
 2   REGION_CODE                  str    
 3   ICB_NAME                     str    
 4   ICB_CODE                     str    
 5   DISPENSER_ACCOUNT_TYPE       str    
 6   BNF_PRESENTATION_CODE        str    
 7   BNF_PRESENTATION_NAME        str    
 8   SNOMED_CODE                  float64
 9   SUPPLIER_NAME                str    
 10  UNIT_OF_MEASURE              str    
 11  GENERIC_BNF_EQUIVALENT_CODE  str    
 12  GENERIC_BNF_EQUIVALENT_NAME  str    
 13  BNF_CHEMICAL_SUBSTANCE_CODE  str    
 14  BNF_CHEMICAL_SUBSTANCE       str    
 15  BNF_PARAGRAPH_CODE           int64  
 16  BNF_PARAGRAPH                str    
 17  BNF_SECTION_CODE             int64  
 18  BNF_SECTION                  str    
 19  BNF_CHAPTER

Inspect the Dataset Structure

General information about the dataset is displayed, including the data type of each column, the number of non-null values, and the overall memory usage. This helps identify potential data quality issues such as missing values and incorrect data types.

2.5 Check missing values

In [32]:
# Count missing values in each column
# This helps us decide how to handle incomplete data
df.isna().sum()

YEAR_MONTH                           0
REGION_NAME                          0
REGION_CODE                          0
ICB_NAME                             0
ICB_CODE                             0
DISPENSER_ACCOUNT_TYPE               0
BNF_PRESENTATION_CODE                0
BNF_PRESENTATION_NAME                0
SNOMED_CODE                       2665
SUPPLIER_NAME                  1426179
UNIT_OF_MEASURE                      0
GENERIC_BNF_EQUIVALENT_CODE          0
GENERIC_BNF_EQUIVALENT_NAME          2
BNF_CHEMICAL_SUBSTANCE_CODE          0
BNF_CHEMICAL_SUBSTANCE               0
BNF_PARAGRAPH_CODE                   0
BNF_PARAGRAPH                        0
BNF_SECTION_CODE                     0
BNF_SECTION                          0
BNF_CHAPTER_CODE                     0
BNF_CHAPTER                          0
PREP_CLASS                           0
PRESCRIBED_PREP_CLASS                0
ITEMS                                0
TOTAL_QUANTITY                       0
NIC                      

Check for Missing Values

The number of missing values in each column is calculated to identify incomplete data and determine whether missing values need to be removed, filled, or investigated further.

2.6 Check duplicates

In [33]:
# Count duplicate rows in the combined dataset
# Duplicate records can affect analysis results
df.duplicated().sum()

np.int64(0)

Check for Duplicate Records

The dataset is examined for duplicate rows to ensure that repeated records do not affect the analysis or produce misleading results.

2.7 Summary statistics

In [34]:
# Generate statistical summaries for numeric columns
# Helps identify ranges, averages, and possible outliers
df.describe()

,YEAR_MONTH,SNOMED_CODE,BNF_PARAGRAPH_CODE,BNF_SECTION_CODE,BNF_CHAPTER_CODE,PREP_CLASS,PRESCRIBED_PREP_CLASS,ITEMS,TOTAL_QUANTITY,NIC
count,6.876045e+06,6.873380e+06,6.876045e+06,6.876045e+06,6.876045e+06,6.876045e+06,6.876045e+06,6.876045e+06,6.876045e+06,6.876045e+06
mean,2.025393e+05,2.262637e+16,4.223487e+04,1.113973e+03,1.102025e+01,3.072261e+00,2.909467e+00,1.892374e+02,1.525446e+04,1.702886e+03
std,4.414698e+01,1.525504e+16,4.118928e+04,7.875419e+02,7.774044e+00,1.180180e+00,1.237379e+00,1.875090e+03,1.994014e+05,1.402299e+04
min,2.025050e+05,2.110000e+11,2.001000e+03,1.010000e+02,1.000000e+00,1.000000e+00,1.000000e+00,1.000000e+00,3.600000e-01,0.000000e+00
25%,2.025070e+05,7.875911e+15,2.305000e+03,4.070000e+02,4.000000e+00,3.000000e+00,2.000000e+00,1.000000e+00,3.000000e+01,3.176000e+01
50%,2.025100e+05,2.176371e+16,4.020100e+04,9.060000e+02,9.000000e+00,3.000000e+00,3.000000e+00,4.000000e+00,1.680000e+02,1.272400e+02
75%,2.026010e+05,3.809981e+16,7.030100e+04,2.101000e+03,2.100000e+01,4.000000e+00,4.000000e+00,2.400000e+01,1.260000e+03,5.726900e+02
max,2.026040e+05,4.586691e+16,1.914020e+05,2.399000e+03,2.300000e+01,5.000000e+00,5.000000e+00,2.525920e+05,1.051406e+08,3.119512e+06


Generate Summary Statistics

Descriptive statistics are generated for the numerical columns to understand their distribution, central tendency, spread, and potential outliers before beginning the cleaning process.

Data Quality Issues Identified

During the initial inspection of the NHS Prescription Cost Analysis dataset, several data quality checks were performed. The following issues were identified and addressed:

- Missing values: The `GENERIC_BNF_EQUIVALENT_NAME` column contained 2 missing values. Since these records could not be reliably categorised, the affected rows were removed.

- Incorrect data type: The `YEAR_MONTH` column was stored in YYYYMM integer format (for example, 202505). This was converted into a datetime format to support time-based analysis and visualisation.

- Duplicate records: Duplicate rows were checked using Pandas' `duplicated()` function. Exact duplicate records were identified and later removed using `drop_duplicates()` to prevent repeated entries from affecting prescription cost and frequency analysis.

- Irrelevant variables: Several administrative and classification columns, such as codes and identifiers, were not required for answering the research questions. These unnecessary columns were removed to simplify the dataset.

Step 3 - Data Cleaning

Clean the combined dataset by handling missing values, removing unnecessary columns, checking data quality, and preparing the data for analysis and visualisation.

3.1 Create a copy of the raw dataset

In [35]:
# Create a copy of the original dataset
# This preserves the raw data for future reference
df_clean = df.copy()

# Display the dimensions of the copied dataset
df_clean.shape

(6876045, 27)

Create a copy of the raw dataset

Before making any modifications, a copy of the original combined dataset is created. This preserves the raw data and ensures that all cleaning operations are performed on a separate dataframe.

3.2 Remove unecessary columns

In [36]:
# Define columns that are not required for the analysis
# These columns contain identifiers, codes, or information outside the project scope
columns_to_drop = [
    "REGION_CODE",
    "ICB_CODE",
    "BNF_PRESENTATION_CODE",
    "SNOMED_CODE",
    "SUPPLIER_NAME",
    "GENERIC_BNF_EQUIVALENT_CODE",
    "BNF_CHEMICAL_SUBSTANCE_CODE",
    "BNF_PARAGRAPH_CODE",
    "BNF_SECTION_CODE",
    "BNF_CHAPTER_CODE",
    "PREP_CLASS",
    "PRESCRIBED_PREP_CLASS",
    "PHARMACY_ADVANCED_SERVICE"
]

# Remove the selected columns from the cleaned dataframe
df_clean = df_clean.drop(columns=columns_to_drop, errors='ignore')

# Display the updated dataset dimensions
df_clean.shape

(6876045, 14)

In [37]:
# Display the remaining columns after removing unnecessary variables
df_clean.columns.tolist()

['YEAR_MONTH',
 'REGION_NAME',
 'ICB_NAME',
 'DISPENSER_ACCOUNT_TYPE',
 'BNF_PRESENTATION_NAME',
 'UNIT_OF_MEASURE',
 'GENERIC_BNF_EQUIVALENT_NAME',
 'BNF_CHEMICAL_SUBSTANCE',
 'BNF_PARAGRAPH',
 'BNF_SECTION',
 'BNF_CHAPTER',
 'ITEMS',
 'TOTAL_QUANTITY',
 'NIC']

Remove Unnecessary Columns

Several columns contained administrative codes, identifiers, and classification fields that were not required for answering the research questions. Removing these variables reduced dataset complexity while retaining the information needed to analyse prescription costs, regional variation, and medicine demand.


3.3 Handle missing values

In [38]:
# Count missing values in each remaining column
# This helps identify columns that require further cleaning
df_clean.isna().sum()

YEAR_MONTH                     0
REGION_NAME                    0
ICB_NAME                       0
DISPENSER_ACCOUNT_TYPE         0
BNF_PRESENTATION_NAME          0
UNIT_OF_MEASURE                0
GENERIC_BNF_EQUIVALENT_NAME    2
BNF_CHEMICAL_SUBSTANCE         0
BNF_PARAGRAPH                  0
BNF_SECTION                    0
BNF_CHAPTER                    0
ITEMS                          0
TOTAL_QUANTITY                 0
NIC                            0
dtype: int64

Handle Missing Values

Missing values were identified in the GENERIC_BNF_EQUIVALENT_NAME column. Since only two records were affected and the missing values could not be recovered, these rows were removed.


In [39]:
# Remove rows where the generic medicine name is missing
# These records cannot be reliably categorised during analysis
df_clean = df_clean.dropna(
    subset=["GENERIC_BNF_EQUIVALENT_NAME"]
)

# Check that missing values have been removed
df_clean.isna().sum()

YEAR_MONTH                     0
REGION_NAME                    0
ICB_NAME                       0
DISPENSER_ACCOUNT_TYPE         0
BNF_PRESENTATION_NAME          0
UNIT_OF_MEASURE                0
GENERIC_BNF_EQUIVALENT_NAME    0
BNF_CHEMICAL_SUBSTANCE         0
BNF_PARAGRAPH                  0
BNF_SECTION                    0
BNF_CHAPTER                    0
ITEMS                          0
TOTAL_QUANTITY                 0
NIC                            0
dtype: int64

Remove Rows With Missing Generic Medicine Names

The `GENERIC_BNF_EQUIVALENT_NAME` column contains only two missing values, representing a negligible proportion of the dataset. Since these records cannot be reliably assigned a generic medicine name, the affected rows are removed to maintain data quality.


3.4 Verify and convert data types (if needed)

In [40]:
# Display the data type of each column
# This helps identify columns that may require conversion
df_clean.dtypes

YEAR_MONTH                       int64
REGION_NAME                        str
ICB_NAME                           str
DISPENSER_ACCOUNT_TYPE             str
BNF_PRESENTATION_NAME              str
UNIT_OF_MEASURE                    str
GENERIC_BNF_EQUIVALENT_NAME        str
BNF_CHEMICAL_SUBSTANCE             str
BNF_PARAGRAPH                      str
BNF_SECTION                        str
BNF_CHAPTER                        str
ITEMS                            int64
TOTAL_QUANTITY                 float64
NIC                            float64
dtype: object

Verify and Convert Data Types

The data types of each column are reviewed to ensure that variables are stored in the correct format for analysis. Columns containing dates, numerical measurements, and categorical information are converted where necessary to improve accuracy and support further analysis.

In [41]:
# Convert YEAR_MONTH from YYYYMM integer format into datetime format
df_clean["YEAR_MONTH"] = pd.to_datetime(
    df_clean["YEAR_MONTH"].astype(str),
    format="%Y%m"
)

# Check the converted values
df_clean["YEAR_MONTH"].head()

0   2025-05-01
1   2025-05-01
2   2025-05-01
3   2025-05-01
4   2025-05-01
Name: YEAR_MONTH, dtype: datetime64[us]

Convert Year-Month Column to Date Format

The YEAR_MONTH column is currently stored as an integer (for example, 202505 represents May 2025). It is converted into a proper date format to make time-based analysis and visualisation easier.

In [42]:
df_clean.head()

,YEAR_MONTH,REGION_NAME,ICB_NAME,DISPENSER_ACCOUNT_TYPE,BNF_PRESENTATION_NAME,UNIT_OF_MEASURE,GENERIC_BNF_EQUIVALENT_NAME,BNF_CHEMICAL_SUBSTANCE,BNF_PARAGRAPH,BNF_SECTION,BNF_CHAPTER,ITEMS,TOTAL_QUANTITY,NIC
0,2025-05-01,MIDLANDS,NHS NORTHAMPTONSHIRE INTEGRATED CARE BOARD,English Dispensing Doctor,diabact UBT 50mg tablets,tablet,Urea [13-C] 50mg tablets breath test kit,Other test for helicobacter pylori preparations,Test for helicobacter pylori,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,1,1.0,21.25
1,2025-05-01,EAST OF ENGLAND,NHS CAMBRIDGESHIRE AND PETERBOROUGH INTEGRATED...,English Pharmacy,diabact UBT 50mg tablets,tablet,Urea [13-C] 50mg tablets breath test kit,Other test for helicobacter pylori preparations,Test for helicobacter pylori,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,1,1.0,21.25
2,2025-05-01,NORTH WEST,NHS GREATER MANCHESTER INTEGRATED CARE BOARD,English Pharmacy,Cimetidine 400mg tablets,tablet,Cimetidine 400mg tablets,Cimetidine,H2-receptor antagonists,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,315,14267.0,1320.12
3,2025-05-01,NORTH EAST AND YORKSHIRE,NHS HUMBER AND NORTH YORKSHIRE INTEGRATED CARE...,English Pharmacy,Cimetidine 400mg tablets,tablet,Cimetidine 400mg tablets,Cimetidine,H2-receptor antagonists,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,146,8061.0,745.76
4,2025-05-01,NORTH EAST AND YORKSHIRE,NHS SOUTH YORKSHIRE INTEGRATED CARE BOARD,English Pharmacy,Tagamet 200mg/5ml syrup,ml,Cimetidine 200mg/5ml oral solution,Cimetidine,H2-receptor antagonists,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,1,1200.0,56.98


3.5 Optional Verification Cell

In [44]:
# Final check of the cleaned dataset
print("Shape:", df_clean.shape)
print("\nMissing values:")
print(df_clean.isna().sum())
print("\nDuplicate rows:", df_clean.duplicated().sum())

Shape: (6876043, 14)

Missing values:
YEAR_MONTH                     0
REGION_NAME                    0
ICB_NAME                       0
DISPENSER_ACCOUNT_TYPE         0
BNF_PRESENTATION_NAME          0
UNIT_OF_MEASURE                0
GENERIC_BNF_EQUIVALENT_NAME    0
BNF_CHEMICAL_SUBSTANCE         0
BNF_PARAGRAPH                  0
BNF_SECTION                    0
BNF_CHAPTER                    0
ITEMS                          0
TOTAL_QUANTITY                 0
NIC                            0
dtype: int64

Duplicate rows: 143869


Final Data Quality Verification Before Duplicate Removal

The cleaned dataset was reviewed to confirm the current data quality status after handling missing values, removing unnecessary columns, and converting data types.

The verification showed that:
- The dataset contains 6,876,043 rows and 14 columns after the initial cleaning steps.
- No missing values remain in any of the retained variables.
- Duplicate records were still present, with 143,869 exact duplicate rows identified. These duplicate records were investigated and removed in the following step to prevent repeated entries from affecting the analysis.

3.6 Investigate duplicate records

In [45]:
# Count the number of duplicate rows in the cleaned dataset
duplicate_count = df_clean.duplicated().sum()

duplicate_count

np.int64(143869)

3.7 View duplicate examples

In [46]:
# Display examples of duplicated records
df_clean[df_clean.duplicated(keep=False)].head(10)

,YEAR_MONTH,REGION_NAME,ICB_NAME,DISPENSER_ACCOUNT_TYPE,BNF_PRESENTATION_NAME,UNIT_OF_MEASURE,GENERIC_BNF_EQUIVALENT_NAME,BNF_CHEMICAL_SUBSTANCE,BNF_PARAGRAPH,BNF_SECTION,BNF_CHAPTER,ITEMS,TOTAL_QUANTITY,NIC
112,2025-05-01,SOUTH EAST,NHS HAMPSHIRE AND ISLE OF WIGHT INTEGRATED CAR...,English Pharmacy,ActiLymph class 1 (18-21mmHg) below knee close...,device,ActiLymph class 1 (18-21mmHg) below knee close...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,4.0,62.32
188,2025-05-01,SOUTH WEST,NHS SOMERSET INTEGRATED CARE BOARD,English Pharmacy,Omeprazole 20mg/5ml oral suspension sugar free,ml,Omeprazole 20mg/5ml oral suspension sugar free,Omeprazole,Proton pump inhibitors,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,1,75.0,234.00
428,2025-05-01,EAST OF ENGLAND,"NHS BEDFORDSHIRE, LUTON AND MILTON KEYNES INTE...",English Dispensing Doctor,Pentasa Mesalazine 1g/100ml enema,enema,Mesalazine 1g/100ml enema,Mesalazine (Systemic),Aminosalicylates,Chronic bowel disorders,Gastro-Intestinal System,1,28.0,70.92
467,2025-05-01,LONDON,NHS NORTH CENTRAL LONDON INTEGRATED CARE BOARD,English Pharmacy,ActiLymph class 1 (18-21mmHg) below knee close...,device,ActiLymph class 1 (18-21mmHg) below knee close...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,2.0,31.16
508,2025-05-01,MIDLANDS,NHS NORTHAMPTONSHIRE INTEGRATED CARE BOARD,English Pharmacy,ActiLymph class 1 (18-21mmHg) below knee close...,device,ActiLymph class 1 (18-21mmHg) below knee close...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,2.0,31.16
509,2025-05-01,SOUTH WEST,"NHS BATH AND NORTH EAST SOMERSET, SWINDON AND ...",English Pharmacy,ActiLymph class 1 (18-21mmHg) below knee close...,device,ActiLymph class 1 (18-21mmHg) below knee close...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,2.0,31.16
556,2025-05-01,EAST OF ENGLAND,NHS HERTFORDSHIRE AND WEST ESSEX INTEGRATED CA...,English Pharmacy,Lansoprazole 15mg gastro-resistant capsules,capsule,Lansoprazole 15mg gastro-resistant capsules,Lansoprazole,Proton pump inhibitors,Antisecretory drugs and mucosal protectants,Gastro-Intestinal System,1,28.0,0.84
649,2025-05-01,MIDLANDS,NHS NORTHAMPTONSHIRE INTEGRATED CARE BOARD,English Pharmacy,ActiLymph class 2 (23-32mmHg) below knee open ...,device,ActiLymph class 2 (23-32mmHg) below knee open ...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,2.0,31.16
652,2025-05-01,SOUTH EAST,NHS HAMPSHIRE AND ISLE OF WIGHT INTEGRATED CAR...,English Pharmacy,Jobst Opaque class 1 (18-21mmHg) below knee cl...,device,Jobst Opaque class 1 (18-21mmHg) below knee cl...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,2.0,29.25
653,2025-05-01,EAST OF ENGLAND,NHS CAMBRIDGESHIRE AND PETERBOROUGH INTEGRATED...,English Pharmacy,Jobst Opaque class 1 (18-21mmHg) below knee cl...,device,Jobst Opaque class 1 (18-21mmHg) below knee cl...,Lymphoedema Garments,Lymphoedema Garments,Lymphoedema Garments,Appliances,1,2.0,29.25


3.8 Check duplicate impact

In [47]:
# Compare dataset size before and after removing duplicates
print("Original cleaned dataset shape:", df_clean.shape)

print("Shape after removing duplicates:", 
      df_clean.drop_duplicates().shape)

Original cleaned dataset shape: (6876043, 14)
Shape after removing duplicates: (6732174, 14)


In [54]:
# Remove exact duplicate rows
df_clean = df_clean.drop_duplicates()

# Confirm duplicates have been removed
print("Duplicate rows after removal:", df_clean.duplicated().sum())

Duplicate rows after removal: 0


In [55]:
# Final verification after duplicate removal
print("Final dataset shape:", df_clean.shape)

print("\nMissing values:")
print(df_clean.isna().sum())

print("\nDuplicate rows:", df_clean.duplicated().sum())

Final dataset shape: (6732174, 14)

Missing values:
YEAR_MONTH                     0
REGION_NAME                    0
ICB_NAME                       0
DISPENSER_ACCOUNT_TYPE         0
BNF_PRESENTATION_NAME          0
UNIT_OF_MEASURE                0
GENERIC_BNF_EQUIVALENT_NAME    0
BNF_CHEMICAL_SUBSTANCE         0
BNF_PARAGRAPH                  0
BNF_SECTION                    0
BNF_CHAPTER                    0
ITEMS                          0
TOTAL_QUANTITY                 0
NIC                            0
dtype: int64

Duplicate rows: 0


Remove Duplicate Records

Since exact duplicate rows were identified during the data quality assessment, duplicate records were removed using Pandas' `drop_duplicates()` function. This ensures that repeated entries do not affect calculations of NHS prescription costs and prescription frequency.

Final Cleaning Outcome

After removing duplicate records, the dataset was verified again. No missing values remained, and all exact duplicate rows were removed. The resulting dataset is now ready for exploratory data analysis and visualisation.

3.9 Save the cleaned dataset

In [56]:
# Save the cleaned dataset as a new CSV file
# index=False prevents pandas from writing the row index as a separate column
df_clean.to_csv("../data/cleaned_nhs_prescription_data.csv", index=False)

# Display a confirmation message
print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


Step 4 — Exploratory Data Analysis (EDA)

Exploratory Data Analysis is performed to understand the structure, patterns, and key characteristics of the cleaned NHS prescription dataset. This stage helps identify important trends and relationships before creating visualisations and drawing conclusions.

4.1 Dataset Overview

The cleaned dataset is reviewed using summary information and descriptive statistics to understand the number of records, available variables, and numerical distributions.


In [70]:
# Display the final dataset shape
df_clean.shape

(6876043, 14)

In [71]:
# Display summary information about columns and data types
df_clean.info()

<class 'pandas.DataFrame'>
Index: 6876043 entries, 0 to 6876044
Data columns (total 14 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   YEAR_MONTH                   datetime64[us]
 1   REGION_NAME                  str           
 2   ICB_NAME                     str           
 3   DISPENSER_ACCOUNT_TYPE       str           
 4   BNF_PRESENTATION_NAME        str           
 5   UNIT_OF_MEASURE              str           
 6   GENERIC_BNF_EQUIVALENT_NAME  str           
 7   BNF_CHEMICAL_SUBSTANCE       str           
 8   BNF_PARAGRAPH                str           
 9   BNF_SECTION                  str           
 10  BNF_CHAPTER                  str           
 11  ITEMS                        int64         
 12  TOTAL_QUANTITY               float64       
 13  NIC                          float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(10)
memory usage: 786.9 MB


In [72]:
# Generate descriptive statistics for numerical columns
df_clean.describe()

,YEAR_MONTH,ITEMS,TOTAL_QUANTITY,NIC
count,6876043,6.876043e+06,6.876043e+06,6.876043e+06
mean,2025-10-14 15:58:32.310496,1.892375e+02,1.525447e+04,1.702886e+03
min,2025-05-01 00:00:00,1.000000e+00,3.600000e-01,0.000000e+00
25%,2025-07-01 00:00:00,1.000000e+00,3.000000e+01,3.176000e+01
50%,2025-10-01 00:00:00,4.000000e+00,1.680000e+02,1.272400e+02
75%,2026-01-01 00:00:00,2.400000e+01,1.260000e+03,5.726900e+02
max,2026-04-01 00:00:00,2.525920e+05,1.051406e+08,3.119512e+06
std,NaN,1.875090e+03,1.994014e+05,1.402299e+04


4.2 Define Analysis Questions

The analysis focuses on understanding NHS prescription patterns by investigating cost, regional variation, and prescription frequency. The following questions guide the exploratory analysis and visualisation process:

1. Which medicines have the highest NHS prescription costs?
2. Which regions have the highest prescription spending?
3. Which medicines are prescribed most frequently?

These questions were selected because they provide insights into NHS expenditure, geographical differences, and medicine demand patterns.

Step 4.3 — Summary statistics for key variables

In [73]:
# Total NHS prescription cost
df_clean["NIC"].sum()

np.float64(11709115113.36)

In [74]:
# Total prescription items
df_clean["ITEMS"].sum()

np.int64(1301204909)

In [75]:
# Number of unique regions
df_clean["REGION_NAME"].nunique()

7

In [76]:
# Number of unique medicines
df_clean["BNF_PRESENTATION_NAME"].nunique()

25978

In [77]:
df_clean["YEAR_MONTH"].nunique()

12

In [78]:
df_clean["YEAR_MONTH"].value_counts().sort_index()

YEAR_MONTH
2025-05-01    579160
2025-06-01    579390
2025-07-01    584548
2025-08-01    569721
2025-09-01    578694
2025-10-01    581500
2025-11-01    571459
2025-12-01    579041
2026-01-01    579687
2026-02-01    570488
2026-03-01    581022
2026-04-01    521333
Name: count, dtype: int64

4.4 Smaller Analysis Tables

Table 1: Top 10 Medicines by NHS Spending (Q1)

Research question: Which medicines have the highest NHS spending?

In [90]:
top_spending_medicines = (
    df_clean.groupby("BNF_PRESENTATION_NAME")["NIC"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
    .reset_index()
)

top_spending_medicines

,BNF_PRESENTATION_NAME,NIC
0,FreeStyle Libre 2 Plus Sensor,3.117698e+08
1,Mounjaro KwikPen 5mg/0.6ml inj 2.4ml pre-fille...,2.442124e+08
2,Dapagliflozin 10mg tablets,1.902574e+08
3,Mounjaro KwikPen 7.5mg/0.6ml inj 2.4ml pre-fil...,1.421275e+08
4,Lixiana 60mg tablets,1.276422e+08
5,Trimbow 87microg/dose / 5microg/dose / 9microg...,1.249804e+08
6,Fostair 100micrograms/dose / 6micrograms/dose ...,1.203291e+08
7,"Adjuvanted trivalent flu vacc (SA, inact) inj ...",1.170129e+08
8,Mounjaro KwikPen 10mg/0.6ml inj 2.4ml pre-fill...,1.049527e+08
9,Jardiance 10mg tablets,9.560658e+07


In [93]:
top_spending_medicines.to_csv(
    "../data/top_spending_medicines.csv",
    index=False
)

Table 2: Spending by Region (Q2)

Research question: Which NHS regions have the highest prescription spending?

In [91]:
regional_spending = (
    df_clean.groupby("REGION_NAME")["NIC"]
    .sum()
    .sort_values(ascending=False)
    .reset_index()
)

regional_spending

,REGION_NAME,NIC
0,MIDLANDS,2.432936e+09
1,NORTH EAST AND YORKSHIRE,2.072929e+09
2,SOUTH EAST,1.691746e+09
3,NORTH WEST,1.654253e+09
4,EAST OF ENGLAND,1.470753e+09
5,LONDON,1.339325e+09
6,SOUTH WEST,1.047174e+09


In [92]:
regional_spending.to_csv(
    "../data/regional_spending.csv",
    index=False
)

Table 3: Top 10 Most Prescribed Medicines (Q3)

Research question: Which medicines are prescribed most frequently?

In [97]:
top_prescribed_medicines = (
    df_clean.groupby("BNF_PRESENTATION_NAME")["ITEMS"]
    .sum()
    .sort_values(ascending=False)
    .head(10)
    .reset_index()
)

top_prescribed_medicines

,BNF_PRESENTATION_NAME,ITEMS
0,Atorvastatin 20mg tablets,40153019
1,Omeprazole 20mg gastro-resistant capsules,31460551
2,Amlodipine 5mg tablets,24099963
3,Lansoprazole 30mg gastro-resistant capsules,21670553
4,Atorvastatin 40mg tablets,20001773
5,Amlodipine 10mg tablets,16414278
6,Aspirin 75mg dispersible tablets,15088717
7,Lansoprazole 15mg gastro-resistant capsules,14881106
8,Metformin 500mg tablets,14695304
9,Paracetamol 500mg tablets,13763349


In [98]:
top_prescribed_medicines.to_csv(
    "../data/top_prescribed_medicines.csv",
    index=False
)

Optional Table 4: Monthly NHS Spending Trend

In [99]:
monthly_spending = (
    df_clean.groupby("YEAR_MONTH")["NIC"]
    .sum()
    .reset_index()
)

monthly_spending

,YEAR_MONTH,NIC
0,2025-05-01,9.536158e+08
1,2025-06-01,9.344203e+08
2,2025-07-01,9.911351e+08
3,2025-08-01,9.010258e+08
4,2025-09-01,9.915961e+08
5,2025-10-01,1.133657e+09
6,2025-11-01,9.569137e+08
7,2025-12-01,1.020691e+09
8,2026-01-01,9.725425e+08
9,2026-02-01,8.972773e+08


In [100]:
monthly_spending.to_csv(
    "../data/monthly_spending.csv",
    index=False
)